### PR TITLE
Implements #8 according to our discussions in PR #10

### LINKED ISSUE #8: Usage of TestDataLoader for different groovy scripts, which use same names for objects of same, or different types is not possible
First of all, thank you for a really great tool!

I see however some limitations. One of them is the inability to use same names for objects (even for different types) from different independent scripts. This is due to the implementation of `EntityBuilder` as singleton. What is the idea behind that?

What speaks against outsourcing of the DSL implementation into a stand-alone delegate object and let `EntityBuilder` just "wire" all together?

E.g. (code snip from `EntityBuilder`):
```groovy
// ...
protected void buildEntities(Reader entityDefinitionReader) {
    DelegatingScript script = createExecutableScriptFromEntityDefinition(entityDefinitionReader)
    script.setDelegate(new EntityBuilderDSL()); // instead of "this"
    script.run()
}
// ...
```

If you like I can create a pull request to discuss a possible implementation.

### LINKED ISSUE #9: Please open the API for other tools
I'm the owner of [dadrus/jpa-unit](https://github.com/dadrus/jpa-unit) and would like to

1. directly use or
2. enable the usage 

of libraries like yours or e.g. [benas/random-beans](https://github.com/benas/random-beans) directly in test methods or through a "setup" annotation (not yet implemented). For that I need the possibility to hook into the loading process of the entities. Unfortunately the implementation of `TestDataLoader` does too much for my scenarios (It does not only load the entities, but does also persist them and manages transactions). I could implement my own _`TestDataLoader`_, but I would have to do that in the `de.triology.blog.testdataloader` package because of the protected visibility of the entire API.

### PR DESCRIPTION
**Pull request by dadrus**
As discussed in #10, here the PR addressing #8.

Following topics are open (will be checked, when done):
- [x] Package structure reflecting the change
- [x] Remove `blog` from package name
- [ ] Enhancement of `README.md` to include the information on the new API, as well as the corresponding _migration guide_
- [x] Actually the `EntityBuilder` class does not implement a builder any more. It solely purpose now is to execute/load the dsl script with entity definitions. The name of the class should reflect this responsibility, e.g. be renaming this class to something like `EntityScriptReader` or `EntityScriptExecutor`
- [x] `EntityBuilderDSL` ist now the actual builder. So I think it is worth to rename this class to `EntityBuilder`.
- [x] Creating an IntelliJ GDSL (counterpart to the eclipse dsd)

If you feel, there are other topics which needs to be addressed before merging this PR, feel free to add them to the above list.

I have closed #10 as this PR is a replacement for it.

When ready, this PR will close #8 and will also close #9

### CODE REVIEW

**Comment by schnatterer** [comment]
OK, I'm done with my review.

To get started, just some personal notes. I found more reasons why this whole thing was so complex to me:

1. Moving from JUnit to Spock *while doing the refactoring*.
  It's not about the framework, just about the combination of refactoring of new test framework at the same time.
  Here's my dilemma: For example you deleted `EntityBuilderTest`. However, I'm convinced that most of the test cases in this class should still be valid regression tests, even though they might have moved to another class.
  But now, if I search for the test case `setProgrammaticallyCreatedDateProperty` for example, it's gone.
  How can I know if this test case is still covered?
  I do not have the time to track if each and every test case is now covered in one of the spock test cases.
  On the other hand, what I can see is that we have code coverage of about 90% which is very good.
  In addition I did a manual test using the version on your branch in one of our projects that heavily relies on test-data-loader. Its tests still pass.
  So I'm just going to trust that everything still works :-/
  What I'd like to say is: In future, when refactoring, please focus on the refactoring, keeping changes to a bare minimum and perform optimizations in a different PR before or after the functional refactoring.
  That way, the changes will be easier to understand and review and we have a higher change of not braking existing functionality by  being able to track the evolution of our test cases.
2. Another point is that you did a lot of renaming e.g. `EntityBuilder.buildEntities()` to `build()` and its the `entityDefinitionReader` parameter to `reader`.
  Well, you obviously like shorter identifiers. I think, this is matter of taste (e.g.`entityDefinitionReader` tells me more than `reader`, whereas I admit `EntityBuilder.buildEntities` contains a certain redundancy) and should only be changed if there is a good reason for it. Otherwise the next committer might change it again. 
  This renaming does not have a connection with the problem the PR tries to solve. It creates a lot more diffs than necessary, making it harder to review.
And again, this distracts our focus on the aim of the PR.

Anyway, as said, changing this again now would probably be asking too much, so let's just leave it as is. Maybe you can consider my thoughts for future changes, though :-)

Here's two things I would like you to change.

1. You removed some methods with pretty expressive method names, such as `EntityBuilder.createExecutableScriptFromEntityDefinition()`  or `EntityBuilder.ensureNameHasNotYetBeenAssigned()` (now in `EntityBuilderDsl`) in favor of putting their content into a calling method (now `EntityBuilder.build()`).
  Why did you do this? This (again) creates more diffs than necessary. Also I can imagine the original author (I'm not the one who wrote it BTW) did this for a reason:
  This adheres to clean code guide lines, such as "a function should do one thing" (`createExecutableScriptFromEntityDefinition()`) and "exception handling is one thing" (`ensureNameHasNotYetBeenAssigned()`). This makes code so much more maintainable.
  Instead we now have one method (`build`) that clearly does more that one thing.
  We should also add a note to clean code to the `CONTRIBUTING.md` ;-)

2. `Demo` -> `TestDataLoaderIT`:
  To me it's a best practice to place the tests in the same package as the class under test.
  Now that `Demo` is no longer a demo but an IT for `TestDataLoader`, it should reside in the same package, IMO. Surefire and failsafe should be able to distinguish between them by their respective default file name patterns (`Test` vs `IT`).
  We could then move `Department` and `User` to the `testentities` package.
  Please also move the `resources` to the new package, they still live in `demo`.

That's it for the results of my review.

As for merging, I'm not quite sure how you plan to go on with the open topics you mentioned.
Will you add more commits now or will you solve them in a new PR?

Once we're done, I'd merge the changes into the develop branch, where we can fix the remaining issues before releasing test-data-loader 1.0
In addition of the open topics you mentioned that would be

- [ ]  Creating an IntelliJ GDSL (counterpart to the eclipse dsd)
- [ ]  Remove `blog` from package name
- [ ]  Automate releasing to maven central
- [ ]  (optional) sonarcloud.io analysis. It will most likely get complicated, because we now have two test folders: `src/main/java` and `src/main/groovy` :-/

**Comment by dadrus** [comment]
Thank you very much for your review.

Please allow me to comment on it.

Answers to your general notes:
1. _JUnit vs. Spock, respectively the removal of `EntityBuilderTest`_: On the one hand the original implementation of the `EntitxBuilder` was split into three parts - the dsl reader, the dsl script and the listener to allow interested parties to receive the corresponding information. I just failed to find something in `EntityBuilderTest` that I could reuse. On the other hand many tests in `EntityBuilderTest` were testing not the actual implementation of the `EntityBuilder`, rather the Groovy language/JVM. Please don't be angry with me, but `setProgrammaticallyCreatedDateProperty` is one of the best examples for it. Because of this I decided to implement new tests which focus on the functionality of the above mentioned classes. This way this part of the code refactoring led to new classes and new tests since the old tests could not be reused.
2. _Renamings_: As you said by yourself the method name already provides a certain context, so that there is no need to repeat it in the name of a parameter. Anyway, this is indeed, matter of taste:). But as you could see, changes of that type were introduced in more or less completely rewritten code only.

Answers to your requests to change:
1. _Removal of pretty expressive method names_: As with _Renamings_ answers (see 2 above) the change happened only in a completely rewritten code. Since the code we are talking about is pretty much new, I've followed the principles, I follow in my projects. These are mainly as follows:
    - use as short but as expressive names/identifiers as possible
    - don't repeat the the context in the name of the identifier (can be considered as the implementation of DRY as well). And if allow me to site uncle Bob, then I am talking about what he's described in "Add Meaningful Context" and "Don't Add Gratious Context"
    - each function should do one thing. This corresponds pretty much to what you've written above. There is however one thing to highlight: the function has to remain readable and maintainable. Taking these concerns into account, a question arises, when a further decomposition or refactoring of a function makes sense. From my own experience it makes sense, when the function is more then ca 15 lines of code, some of the functionality of a function can be reused in other functions (like e.g. verification of a condition) or when it is obviously that the function makes more than one thing. This approach contradicts a clean code requirement on a function to be small on the one hand, on the other hand it does not scatter the actual functionality across multiple places and thus makes refactoring more difficult. You see, it is kind of a trade-off with a simple goal to use the time effectively and to avoid efforts which on their own could lead to further efforts.

With that in mind let us look at the methods, you've referenced in your comment:
- `EntityBuilder.createExecutableScriptFromEntityDefinition()` - could not be reused. The new code is a different one. 
- `EntityBuilder.ensureNameHasNotYetBeenAssigned()` - could be reused. But it would be referenced in one place only and consists of actually two lines (`if` and `throw`). So I used these two lines directly. The readability and the maintainability do not suffer since one can see the purpose directly.

2. Fully agree. But since `Demo` was residing in the `demo` package, I renamed the package name as well to give it a more meaningful name. There however two good reasons, not to place IT tests in the same package as the CUT(s) (Both are however project and test scope related.):
    - To avoid usage of package scoped definitions in the CUT(s) from the test (IT are considered blck box tests)
    - To be able to run all the integration and unit tests from the IDE independently. Because of this I find it pretty helpful to have an additional test source folder, e.g. integrationtest in which all the IT tests reside. Maybe IntelliJ brings a better support for that type of usability.

**Comment by dadrus** [comment]
Sory, I hit a wrong button:)

**Comment by schnatterer** [comment]
I agree with the list.  

The additional points  in my last comment were rather FYI. To point out what we plan to do between this PR and the 1.0 release. 
However, you're right. The need for the GDSL arises from this PR, so it would be a good idea to fix it in this content. I can see why it's difficult for you to implement the GDSL. Unfortunately, I'm not quite sure how much more time I can invest into the test-data-loader right now. 
Lets decide about that later. If we can't solve it in this PR, we'll just create an issue for it, so this PR does not become a never ending story :-)

So, no further concerns from my side. Please continue with your work.

**Comment by dadrus** [comment]
Hi @schnatterer 

Please check the above commits, especially the last one. Even the gdsl file implements the pattern from [here](http://www.tothenew.com/blog/gdsl-awesomeness-delegating-closure-calls/), the referenced type to the `create` method is not used. Thus auto-completion for its members is not working. Any ideas?

**Comment by dadrus** [comment]
Only the `README.md` is open from the above list. Do you have any concerns regarding the above commits? If these are ok for you, I'll see what I can do for the contents of the `README.md` file

File: src/main/groovy/de/triology/testdata/loader/TestDataLoader.groovy [hunk 1]
```diff
@@ -119,14 +119,14 @@ class TestDataLoader implements EntityBuilderListener {
      */
     void loadTestData(Collection<String> entityDefinitionFiles) {
         EntityPersister persister = new EntityPersister(entityManager)
-        EntityBuilder entityBuilder = new EntityBuilder()
+        EntitiesScriptExecutor entityBuilder = new EntitiesScriptExecutor()
```
Comment (schnatterer) [inline src/main/groovy/de/triology/testdata/loader/TestDataLoader.groovy]: Incomplete refactoring :-P

Comment (dadrus) [reply src/main/groovy/de/triology/testdata/loader/TestDataLoader.groovy]: done in a5d89aa :)

**Comment by schnatterer** [comment]
Hi @dadrus,

you created a GDSL! Awesome work! Thanks for your dedication to this project. This is really remarkable, especially when taking into account that you don't use IntelliJ :-)
The completion with GDSL really works amazingly. I didn't have to do anything for it. I'm also surprised that it is only 4 lines of code.

I reviewed the changes. After the variable name change they all look good to me.

As for the README, I can think of tree topics

* How to use the autocompletion feature (dsdl/GDSL) - for me it worked just like that, so no need to describe it in the README, IMO. How about eclipse?
* The implementation as described in the original article mentioned in the README is now different.
  In order to avoid confusion, my idea would be to reflect this fact in a short sentence such as
  *Please note that from version 1.x the implementation as described in the article has now changed. For example, the functionality of `EntityBuilder` from the article is now realized in `EntitiesScriptExecutor` and `EntityBuilder`. See #10 and #12 for details.*
* Migration. My idea would be to add a paragraph in the README (maybe before *Tested with...*) such as the following (please review, extend, improve, etc.):


# Migrating
## From 0.x to 1.x

* In the tests: Change your imports from `de.triology.blog.testdataloader.TestDataLoader` to `de.triology.testdata.loader.TestDataLoader`
* In your data groovy scripts: Just remove `import static de.triology.blog.testdataloader.EntityBuilder.create` statements. 
  Auto completion is now realized using DSL Descriptors (DSLD) for eclipse and GDSL


That should do it. What do you think?

**Comment by dadrus** [comment]
The commit above includes the minimal `README.md` changes. It mainly contains fixes for misspelled words, like _entitiy_, which became _entity_, a sentense which more or less corresponds to the second topic you proposed in your comment and package name changes for the classes used by the integration test.

To be honest, I think it would be better if the API change and the corresponding "migration" path would be described in the release note of the coming release. Otherwise you'll end up with more complex project description instead of a description reflecting the current status. What do you think about the following approach:
- remove the references to classes used in the project.
- Instead one can define some classes for the purpose of the usage description in the `README.md` itself. This way possible future API changes (I think for example about a more fluent DSL API) will not be spread across the entire description.
- Introduce a section, which describes the main API of the library. As of today this information is partially present but also scattered in the `README.md`.
- Add release notes to each coming release, where all the changes worth to note are described. Like e.g. the migration section from your last comment, API changes etc.
That way the project description will represent the latest development status and will stay clear and also easier to follow for new library users.

If you like, I can change the `README.md` according to the above ideas.

Regarding your question abou dsl descriptors: for both eclipse and IntelliJ the corresponding file has to be present in the classpath. In case of eclipse, there's however a special requirement: the dsld file has to reside in a folder named `dsld`. That's all. The implemented autocompletion for eclipse is somewhat limited compared to the IntelliJ one (I mean the code in dsld, respectively gdsl files). In IntelliJ one just can hit configured keys for autocompletion and all the properties of the class to be built are shown. In eclipse, one have either start typing the property name, or use `this` followed by a dot to see all the properties and functions of the entity class. Maybe this is a limitation of eclipse, maybe of the dsld implementation I did. Presumably the latter:)

**Comment by schnatterer** [comment]
You're right about the migration path. I will just append what I wrote above on the [releases page](https://github.com/triologygmbh/test-data-loader/releases) once the release is there.

As for your ideas about the `README.md`, if you're still motivated to do it, please go ahead :-)
IMO, we can do this in a separate PR, as these improvements are not strictly related to the changes done in this PR.

Please note that I created a `develop` branch, which is now default.
FYI In the meantime I started working on the maven central release which might still happen this year.

🍾 You did it. The PR is merged 🍾
Pleasure to work with you and thanks once more for your endurance and your extensive work. I hope you are now able to embed test-data-loader in dadrus/jpa-unit as described in #9. If there's anything else to be done in this regard, I'm always happy to review more PRs. ☺️

**Comment by dadrus** [comment]
You're very welcome:) The pleasure is on my side as well.

I'll submit a new PR for the `README.md` contents. If you are going to like it - cool. Otherwise - it is your project:)

The changes in this PR are fully sufficient to enable the integral usage of test-data-loader in my project. Now I have to find a good way to do it:) You know the agony of choice between the multiple options:)

**Pull request merged**